lms/services/canvas_api/authenticated.py

```diff
@@ -0,0 +1,140 @@
+        """
+        Send a Canvas API request, and retry it if there are OAuth problems.
+
+        :param method: HTTP method to use (e.g. "POST")
+        :param path: The path in the API to make a request to
+        :param schema: Schema to apply to the return values
+        :param params: Any query parameters to add to the request
+        :raise CanvasAPIAccessTokenError: if the request fails because our
+             Canvas API access token for the user is missing, expired, or has
+             been deleted
+        :return: JSON deserialised object
+        :raise CanvasAPIAccessTokenError: If a token is required and cannot be
+                                          found / refreshed
+        """
+        call_args = (method, path, schema, params)
+
+        try:
+            auth_header = f"Bearer {self._token_store.get().access_token}"
+            return self._api.send(*call_args, headers={"Authorization": auth_header})
+
+        except CanvasAPIAccessTokenError:
+            refresh_token = self._token_store.get().refresh_token
+            if not refresh_token:
+                raise
+
+            auth_header = f"Bearer {self.get_refreshed_token(refresh_token)}"
+            return self._api.send(*call_args, headers={"Authorization": auth_header})
+
+    def get_token(self, authorization_code):
+        """
+        Get an access token for the current LTI user.
+
+        :param authorization_code: The Canvas API OAuth 2.0 authorization code
+                                   to exchange for an access token
+        :return: An access token string
+        """
+        # For documentation of this request see:
+        # https://canvas.instructure.com/doc/api/file.oauth_endpoints.html#post-login-oauth2-token
+        return self._send_token_request(
+            grant_type="authorization_code",
+            code=authorization_code,
+            redirect_uri=self._redirect_uri,
+            replace_tokens=True,
+        )
+
+    def get_refreshed_token(self, refresh_token):
+        """
+        Get a refreshed access token for the current LTI user.
+
+        :param refresh_token: The Canvas API OAuth 2.0 refresh token from a
+                              previous token call
+        :return: A new access token string
+        """
+        return self._send_token_request(
+            grant_type="refresh_token", refresh_token=refresh_token
+        )
+
+    def _send_token_request(self, grant_type, refresh_token=None, **kwargs):
+        params = {
+            "grant_type": grant_type,
+            "client_id": self._client_id,
+            "client_secret": self._client_secret,
+            **kwargs,
+        }
+
+        if refresh_token:
+            params["refresh_token"] = refresh_token
+
+        parsed_params = self._api.send(
+            "POST",
+            "login/oauth2/token",
+            url_stub="",
+            params=params,
+            schema=CanvasTokenResponseSchema,
+        )
```

Comment:
This part is massively simplified by adding the `url_stub` as an argument. This is where all the nonsense was before.